blogposts/2020/evolution-of-the-precise-code-intel-backend.md

```diff
@@ -0,0 +1,183 @@
+
+We decided to _slowly_ rewrite each of the services into Go. Our high-level plan was to first extract the CPU-bound work done by the lsif-server into a high-performance Go process which could be called by the existing worker code (this is similar to the [strangler fig](https://martinfowler.com/bliki/StranglerFigApplication.html) pattern of refactoring). This part of the stack seemed to benefit the most from a rewrite, and allowed us to use existing knowledge to aggressively optimize Go rather than having to learn low-level optimization techniques for Node.js environments. Eventually, more and more of the worker would fall into the Go code's event horizon, and eventually the entire worker would be replaced.
+
+But actions don't always occur as they're planned.
+
+Turns out it's pretty easy to write the same system in a language you know well after fixing all the bugs in a different language, and I just rewrote all three services in a single pass. The resulting code wasn't particularly idiomatic as it really was meant to be a fast, behaviorally-equivalent translation into the target language. Constant, small refactorings have been occurring since the rewrite in order to shape the code into something more appropriate for the new environment.
+
+This has unlocked a large number of performance improvement opportunities, the results of which will be detailed in the Sourcegraph 3.17 release announcement. The technical details are described in a blog post on [performance improvements for precise code intel](/blog/performance-improvements-in-precise-code-intel).
+
+## Remove the API server
+
+**PRs**: 
+- [codeintel: Call handleEnqueue from lsifserver proxy (#10871)](https://github.com/sourcegraph/sourcegraph/pull/10871)
+- [codeintel: Call query methods from from lsifserver client (#10872)](https://github.com/sourcegraph/sourcegraph/pull/10872)
+- [Remove precise-code-intel-api-server service (#10906)](https://github.com/sourcegraph/sourcegraph/pull/10906)
+
+After rewriting the services in Go, the lsif-server process maintained no local state, and queries to it no longer crossed a language boundary. There was little reason to keep the service separate. The client used by the frontend's HTTP API and GraphQL resolvers was able to directly perform the same functions as the old server's HTTP handlers, and the newly unused HTTP server/routing code could be dropped. 
+
+![architecture diagram](https://storage.googleapis.com/sourcegraph-assets/lsif-arch-8.png)
+
+This change is discussed in more detail in a blog post on [performance improvements for precise code intel](/blog/performance-improvements-in-precise-code-intel/#collapse-network-boundaries).
+
+<style>
```

Comment:
Love this post! Have you considered adding any sort of wrap-up at the end? It felt a bit sudden. Not critical, just something I noticed

Reply:
Yeah, I'd like to leave something to end on, but I'm not sure what the most beneficial thing to reiterate would be. Any suggestions?

Reply:
I'd love to hear what this is all building to — a roadmap for the future! E.g., describe how all of this is in service of enabling precise code intel on _massive_ deployments, and ultimately, on Sourcegraph Cloud. E.g. describe the auto-indexing work, etc.